### ISSUE TITLE
Fix .stylelintcache should be refreshed when config is changed

### ISSUE THREAD

**Issue by cmalard** [stylelint/stylelint#2908]
Hello there,

A config change is not detected when using cache. It should be refreshed.

> Which version of stylelint are you using?

`8.1.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

`stylelint  **/*.css --cache`

**Comment by evilebottnawi** [comment]
confirmed

**Comment by jeddy3** [comment]
@cmalard Thanks for the report and for using the template. @evilebottnawi Has correctly label this as "help wanted". Feel free to dig and contribute a PR if you'd like to see this fixed.

**Comment by cmalard** [comment]
ESLint is doing it, it could help : 
https://github.com/eslint/eslint/blob/9d1df92628dd4dd1e70fbb19454008e146387435/lib/cli-engine.js#L507-L509

**Comment by xarety** [comment]
Hey guys, any updates here?

**Comment by ybiquitous** [comment]
@xarety Nobody touches this issue now. Please consider [contributing](https://stylelint.io/contributing) if you have time!

**Comment by ybiquitous** [comment]
I tried looking into this problem a bit.

The `standalone()` function calculates a hash of a given config object here:

https://github.com/stylelint/stylelint/blob/592b149a6225559cccada30d313814910439d170/lib/standalone.js#L179-L182

When using directly `standalone()` (that is, the Node.js API), the `config` variable has a non-nullable value. On the other hand, when using CLI, `config` is nullable. Because a config file (e.g. `.styelintrc` or a `—config` argument) is loaded later: in the `lintSource()` function

https://github.com/stylelint/stylelint/blob/592b149a6225559cccada30d313814910439d170/lib/standalone.js#L228-L230

https://github.com/stylelint/stylelint/blob/592b149a6225559cccada30d313814910439d170/lib/lintSource.js#L54-L60

In summary, I think we need to move the location of hash calculation for `config`; maybe into `lintSource()`.

**Comment by jeddy3** [comment]
@ybiquitous Thanks for digging into this issue!

> In summary, I think we need to move the location of hash calculation for config; maybe into lintSource().

Sounds like an appropriate change to me.

**Comment by kimulaco** [comment]
I would like to work on this issue.
